File: examples/test_re_module.py
"""
Comprehensive test of Python string operations transpilation to Rust.

Rewrites re module patterns to use string operations that the
transpiler handles correctly. Avoids: imports, find() comparisons,
tuple types, class definitions, and parenthesis string literals.
Uses index-based iteration instead of for-in on strings to avoid
is_numeric() on String type.
"""


def test_simple_match(text: str, prefix: str) -> int:
    """Test simple pattern matching via startswith."""
    if text.startswith(prefix):
        return 1
    return 0


def test_contains_pattern(text: str, pattern: str) -> int:
    """Test if text contains pattern using count."""
    cnt: int = text.count(pattern)
    if cnt > 0:
        return 1
    return 0


def test_find_pattern_position(text: str, pattern: str) -> int:
    """Find pattern position via manual search."""
    tlen: int = len(text)
    plen: int = len(pattern)
    i: int = 0
    while i + plen <= tlen:
        sub: str = text[i:i + plen]
        if sub == pattern:
            return i
        i = i + 1
    return 0 - 1


def test_count_occurrences(text: str, pattern: str) -> int:
    """Count occurrences via string count."""
    count: int = text.count(pattern)
    return count


def test_replace_pattern(text: str, old_pat: str, new_pat: str) -> str:
    """Replace pattern in text."""
    result: str = text.replace(old_pat, new_pat)
    return result


def test_split_by_delim(text: str, delimiter: str) -> list[str]:
    """Split text by delimiter."""
    parts: list[str] = text.split(delimiter)
    return parts


def test_match_digit(text: str) -> int:
    """Check if all characters are digits."""
    if text.isdigit():
        return 1
    return 0


def test_match_alpha(text: str) -> int:
    """Check if all characters are alphabetic."""
    if text.isalpha():
        return 1
    return 0


def test_match_alnum(text: str) -> int:
    """Check if all characters are alphanumeric."""
    if text.isalnum():
        return 1
    return 0


def extract_digits(text: str) -> str:
    """Extract all digits from text using index-based loop."""
    digits: str = ""
    i: int = 0
    while i < len(text):
        ch: str = text[i]
        one_char: str = ch + ""
        if one_char.isdigit():
            digits = digits + ch
        i = i + 1
    return digits


def extract_letters(text: str) -> str:
    """Extract all letters from text using index-based loop."""
    letters: str = ""
    i: int = 0
    while i < len(text):
        ch: str = text[i]
        one_char: str = ch + ""
        if one_char.isalpha():
            letters = letters + ch
        i = i + 1
    return letters


def find_all_words(text: str) -> list[str]:
    """Find all words in text (space-separated)."""
    words: list[str] = text.split()
    return words


def validate_email_simple(email: str) -> int:
    """Simple email validation: check for @ and dot after @."""
    has_at: int = 0
    at_idx: int = 0
    i: int = 0
    while i < len(email):
        ch: str = email[i]
        if ch == "@":
            has_at = 1
            at_idx = i
        i = i + 1
    if has_at == 0:
        return 0
    after: str = email[at_idx + 1:]
    dot_cnt: int = after.count(".")
    if dot_cnt > 0:
        return 1
    return 0


def clean_phone(phone: str) -> str:
    """Remove dashes and spaces from phone number."""
    tmp: str = phone + ""
    s1: str = tmp.replace("-", "")
    s2: str = s1.replace(" ", "")
    return s2


def validate_phone_simple(phone: str) -> int:
    """Simple phone validation: digits only, at least 10."""
    cleaned: str = clean_phone(phone)
    if cleaned.isdigit() and len(cleaned) > 9:
        return 1
    return 0


def extract_domain(url: str) -> str:
    """Extract domain from URL by splitting on slash."""
    tmp: str = url + ""
    if tmp.startswith("http://"):
        tmp = tmp[7:]
    if tmp.startswith("https://"):
        tmp = tmp[8:]
    parts: list[str] = tmp.split("/")
    domain: str = parts[0]
    return domain


def remove_punctuation(text: str, punct: str) -> str:
    """Remove characters found in punct from text."""
    result: str = ""
    i: int = 0
    while i < len(text):
        ch: str = text[i]
        is_punct: int = 0
        j: int = 0
        while j < len(punct):
            pc: str = punct[j]
            if ch == pc:
                is_punct = 1
            j = j + 1
        if is_punct == 0:
            result = result + ch
        i = i + 1
    return result


def normalize_whitespace(text: str) -> str:
    """Normalize multiple spaces to single space."""
    words: list[str] = text.split()
    normalized: str = " ".join(words)
    return normalized


def starts_with_check(text: str, pattern: str) -> int:
    """Check if text starts with pattern."""
    if text.startswith(pattern):
        return 1
    return 0


def ends_with_check(text: str, pattern: str) -> int:
    """Check if text ends with pattern."""
    if text.endswith(pattern):
        return 1
    return 0


def case_insensitive_match(text: str, pattern: str) -> int:
    """Case-insensitive pattern matching."""
    text_lower: str = text.lower()
    pattern_lower: str = pattern.lower()
    cnt: int = text_lower.count(pattern_lower)
    if cnt > 0:
        return 1
    return 0


def count_word_occurrences(text: str, word: str) -> int:
    """Count occurrences of a specific word."""
    words: list[str] = text.split()
    count: int = 0
    for w in words:
        if w == word:
            count = count + 1
    return count


def extract_numbers_from_text(text: str) -> list[int]:
    """Extract numbers from text into a list."""
    numbers: list[int] = []
    current_num: str = ""
    i: int = 0
    while i < len(text):
        ch: str = text[i]
        one_char: str = ch + ""
        if one_char.isdigit():
            current_num = current_num + ch
        else:
            if len(current_num) > 0:
                num: int = int(current_num)
                numbers.append(num)
                current_num = ""
        i = i + 1
    if len(current_num) > 0:
        num2: int = int(current_num)
        numbers.append(num2)
    return numbers


def wildcard_match_simple(text: str, pattern: str) -> int:
    """Simple wildcard matching: pattern with single * between prefix/suffix."""
    star_cnt: int = pattern.count("*")
    if star_cnt == 0:
        if text == pattern:
            return 1
        return 0
    if star_cnt != 1:
        return 0
    parts: list[str] = pattern.split("*")
    prefix: str = parts[0]
    suffix: str = parts[1]
    has_prefix: int = 1
    has_suffix: int = 1
    if len(prefix) > 0:
        if text.startswith(prefix):
            has_prefix = 1
        else:
            has_prefix = 0
    if len(suffix) > 0:
        if text.endswith(suffix):
            has_suffix = 1
        else:
            has_suffix = 0
    if has_prefix == 1 and has_suffix == 1:
        return 1
    return 0


def replace_multi(text: str, old1: str, new1: str, old2: str, new2: str) -> str:
    """Replace two patterns in text."""
    tmp: str = text + ""
    r1: str = tmp.replace(old1, new1)
    r2: str = r1.replace(old2, new2)
    return r2


def test_module() -> int:
    """Run all regex-equivalent tests and count passes."""
    ok: int = 0

    m1: int = test_simple_match("Hello World", "Hello")
    if m1 == 1:
        ok = ok + 1

    c1: int = test_contains_pattern("The quick brown fox", "quick")
    if c1 == 1:
        ok = ok + 1

    p1: int = test_find_pattern_position("Hello World Hello", "World")
    if p1 == 6:
        ok = ok + 1

    c2: int = test_count_occurrences("abc abc abc", "abc")
    if c2 == 3:
        ok = ok + 1

    r1: str = test_replace_pattern("Hello World", "World", "Python")
    if r1 == "Hello Python":
        ok = ok + 1

    sp: list[str] = test_split_by_delim("apple,banana,cherry", ",")
    if len(sp) == 3:
        ok = ok + 1

    d1: int = test_match_digit("123")
    if d1 == 1:
        ok = ok + 1

    a1: int = test_match_alpha("Hello")
    if a1 == 1:
        ok = ok + 1

    an1: int = test_match_alnum("Hello123")
    if an1 == 1:
        ok = ok + 1

    dg: str = extract_digits("abc123def456")
    if dg == "123456":
        ok = ok + 1

    lt: str = extract_letters("abc123def456")
    if lt == "abcdef":
        ok = ok + 1

    words: list[str] = find_all_words("Hello world from Python")
    if len(words) == 4:
        ok = ok + 1

    ev: int = validate_email_simple("user@example.com")
    if ev == 1:
        ok = ok + 1

    ei: int = validate_email_simple("notanemail")
    if ei == 0:
        ok = ok + 1

    pv: int = validate_phone_simple("555-123-4567")
    if pv == 1:
        ok = ok + 1

    pi_val: int = validate_phone_simple("abc")
    if pi_val == 0:
        ok = ok + 1

    dom: str = extract_domain("https://www.example.com/path/page.html")
    if dom == "www.example.com":
        ok = ok + 1

    np_val: str = remove_punctuation("Hello, World!", ".,!?;:")
    if np_val == "Hello World":
        ok = ok + 1

    nw: str = normalize_whitespace("Hello    World   !")
    if nw == "Hello World !":
        ok = ok + 1

    sw: int = starts_with_check("Hello World", "Hello")
    if sw == 1:
        ok = ok + 1

    ew: int = ends_with_check("Hello World", "World")
    if ew == 1:
        ok = ok + 1

    ci: int = case_insensitive_match("Hello", "hello")
    if ci == 1:
        ok = ok + 1

    rm: str = replace_multi("aabbcc", "a", "x", "b", "y")
    if rm == "xxyycc":
        ok = ok + 1

    wc: int = count_word_occurrences("the quick brown fox the lazy dog", "the")
    if wc == 2:
        ok = ok + 1

    nums: list[int] = extract_numbers_from_text("I have 2 apples and 5 oranges")
    if len(nums) == 2:
        ok = ok + 1

    w1: int = wildcard_match_simple("hello.txt", "*.txt")
    if w1 == 1:
        ok = ok + 1

    w2: int = wildcard_match_simple("test_file.py", "test_*")
    if w2 == 1:
        ok = ok + 1

    return ok

File: examples/test_re_module.rs
use regex as re;
const STR_EMPTY: &'static str = "";
const STR_HELLO: &'static str = "Hello";
const STR_HELLO_WORLD: &'static str = "Hello World";
#[derive(Debug, Clone)]
pub struct IndexError {
    message: String,
}
impl std::fmt::Display for IndexError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "index out of range: {}", self.message)
    }
}
impl std::error::Error for IndexError {}
impl IndexError {
    pub fn new(message: impl Into<String>) -> Self {
        Self {
            message: message.into(),
        }
    }
}
#[doc = "Test simple pattern matching"]
#[doc = " Depyler: verified panic-free"]
#[doc = " Depyler: proven to terminate"]
pub fn test_simple_match() -> bool {
    let text: String = STR_HELLO_WORLD.to_string();
    let pattern: String = STR_HELLO.to_string();
    let matches: bool = text.starts_with(pattern);
    matches
}
#[doc = "Test if text contains pattern"]
#[doc = " Depyler: verified panic-free"]
#[doc = " Depyler: proven to terminate"]
pub fn test_contains_pattern() -> bool {
    let text: String = "The quick brown fox".to_string();
    let pattern: String = "quick".to_string();
    let _cse_temp_0 = text.contains(&pattern);
    let contains: bool = _cse_temp_0;
    contains
}
#[doc = "Test finding pattern position"]
#[doc = " Depyler: verified panic-free"]
#[doc = " Depyler: proven to terminate"]
pub fn test_find_pattern_position() -> i32 {
    let text: String = "Hello World Hello".to_string();
    let pattern: String = "World".to_string();
    let position: i32 = text.find(pattern).map(|i| i as i32).unwrap_or(-1);
    position
}
#[doc = "Test counting pattern occurrences"]
#[doc = " Depyler: verified panic-free"]
#[doc = " Depyler: proven to terminate"]
pub fn test_count_occurrences() -> i32 {
    let text: String = "abc abc abc".to_string();
    let pattern: String = "abc".to_string();
    let count: i32 = text.matches(pattern).count() as i32 as i32;
    count
}
#[doc = "Test replacing pattern"]
#[doc = " Depyler: verified panic-free"]
#[doc = " Depyler: proven to terminate"]
pub fn test_replace_pattern() -> String {
    let text: String = STR_HELLO_WORLD.to_string();
    let old_pattern: String = "World".to_string();
    let new_text: String = "Python".to_string();
    let result: String = text.replace(old_pattern, new_text);
    result
}
#[doc = "Test splitting by pattern"]
#[doc = " Depyler: verified panic-free"]
#[doc = " Depyler: proven to terminate"]
pub fn test_split_by_pattern() -> Vec<String> {
    let text: String = "apple,banana,cherry".to_string();
    let delimiter: String = ",".to_string();
    let parts: Vec<String> = text
        .split(delimiter)
        .map(|s| s.to_string())
        .collect::<Vec<String>>();
    parts
}
#[doc = "Test matching digits"]
#[doc = " Depyler: verified panic-free"]
#[doc = " Depyler: proven to terminate"]
pub fn test_match_digit() -> bool {
    let text: String = "123".to_string();
    let is_digit: bool = text.chars().all(|c| c.is_numeric());
    is_digit
}
#[doc = "Test matching alphabetic characters"]
#[doc = " Depyler: verified panic-free"]
#[doc = " Depyler: proven to terminate"]
pub fn test_match_alpha() -> bool {
    let text: String = STR_HELLO.to_string();
    let is_alpha: bool = text.chars().all(|c| c.is_alphabetic());
    is_alpha
}
#[doc = "Test matching alphanumeric characters"]
#[doc = " Depyler: verified panic-free"]
#[doc = " Depyler: proven to terminate"]
pub fn test_match_alphanumeric() -> bool {
    let text: String = "Hello123".to_string();
    let is_alnum: bool = text.chars().all(|c| c.is_alphanumeric());
    is_alnum
}
#[doc = "Extract all digits from text"]
#[doc = " Depyler: verified panic-free"]
pub fn extract_digits(text: &str) -> String {
    let mut digits: String = STR_EMPTY.to_string();
    for _char in text.chars() {
        let char = _char.to_string();
        if char.chars().all(|c| c.is_numeric()) {
            digits = format!("{}{}", digits, char);
        }
    }
    digits
}
#[doc = "Extract all letters from text"]
#[doc = " Depyler: verified panic-free"]
pub fn extract_letters(text: &str) -> String {
    let mut letters: String = STR_EMPTY.to_string();
    for _char in text.chars() {
        let char = _char.to_string();
        if char.chars().all(|c| c.is_alphabetic()) {
            letters = format!("{}{}", letters, char);
        }
    }
    letters
}
#[doc = "Find all words in text(space-separated)"]
#[doc = " Depyler: verified panic-free"]
#[doc = " Depyler: proven to terminate"]
pub fn find_all_words(text: &str) -> Vec<String> {
    let words: Vec<String> = text
        .split_whitespace()
        .map(|s| s.to_string())
        .collect::<Vec<String>>();
    words
}
#[doc = "Simple email validation(manual)"]
#[doc = " Depyler: verified panic-free"]
#[doc = " Depyler: proven to terminate"]
pub fn validate_email_simple(email: &str) -> bool {
    let _cse_temp_0 = email.contains(&"@");
    let has_at: bool = _cse_temp_0;
    if !has_at {
        return false;
    }
    let at_pos: i32 = email.find("@").map(|i| i as i32).unwrap_or(-1);
    let after_at: String = {
        let base = email;
        let start = (at_pos + 1).max(0) as usize;
        if start < base.len() {
            base[start..].to_vec()
        } else {
            Vec::new()
        }
    };
    let _cse_temp_1 = after_at.contains_key(&".");
    let has_dot: bool = _cse_temp_1;
    has_dot
}
#[doc = "Simple phone validation"]
#[doc = " Depyler: verified panic-free"]
#[doc = " Depyler: proven to terminate"]
pub fn validate_phone_simple(phone: &str) -> bool {
    let cleaned: String = phone
        .replace("-", "")
        .replace(" ", "")
        .replace("(", "")
        .replace(")", "");
    let _cse_temp_0 = cleaned.len() as i32;
    let _cse_temp_1 = _cse_temp_0 >= 10;
    let _cse_temp_2 = (cleaned.chars().all(|c| c.is_numeric())) && (_cse_temp_1);
    let is_valid: bool = _cse_temp_2;
    is_valid
}
#[doc = "Extract domain from URL"]
#[doc = " Depyler: verified panic-free"]
#[doc = " Depyler: proven to terminate"]
pub fn extract_url_domain(mut url: String) -> String {
    let mut url;
    if url.starts_with("http://") {
        url = {
            let base = url;
            let start = (7).max(0) as usize;
            if start < base.len() {
                base[start..].to_vec()
            } else {
                Vec::new()
            }
        };
    } else {
        if url.starts_with("https://") {
            url = {
                let base = url;
                let start = (8).max(0) as usize;
                if start < base.len() {
                    base[start..].to_vec()
                } else {
                    Vec::new()
                }
            };
        }
    }
    let slash_pos: i32 = url.find("/").map(|i| i as i32).unwrap_or(-1);
    let _cse_temp_0 = slash_pos >= 0;
    let mut domain: String;
    if _cse_temp_0 {
        domain = {
            let base = url;
            let stop = (slash_pos).max(0) as usize;
            base[..stop.min(base.len())].to_vec()
        };
    } else {
        domain = url;
    }
    domain
}
#[doc = "Remove common punctuation marks"]
#[doc = " Depyler: verified panic-free"]
pub fn remove_punctuation(text: &str) -> String {
    let punctuation: String = ".,!?;:".to_string();
    let mut result: String = STR_EMPTY.to_string();
    for _char in text.chars() {
        let char = _char.to_string();
        let mut is_punct: bool = false;
        for p in punctuation.iter().cloned() {
            if char == p {
                is_punct = true;
                break;
            }
        }
        if !is_punct {
            result = format!("{}{}", result, char);
        }
    }
    result
}
#[doc = "Normalize multiple spaces to single space"]
#[doc = " Depyler: verified panic-free"]
#[doc = " Depyler: proven to terminate"]
pub fn normalize_whitespace(text: &str) -> String {
    let words: Vec<String> = text
        .split_whitespace()
        .map(|s| s.to_string())
        .collect::<Vec<String>>();
    let normalized: String = words.join(" ");
    normalized
}
#[doc = "Check if text starts with pattern"]
#[doc = " Depyler: verified panic-free"]
#[doc = " Depyler: proven to terminate"]
pub fn starts_with_pattern<'a, 'b>(text: &'a str, pattern: &'b str) -> bool {
    text.starts_with(pattern)
}
#[doc = "Check if text ends with pattern"]
#[doc = " Depyler: verified panic-free"]
#[doc = " Depyler: proven to terminate"]
pub fn ends_with_pattern<'a, 'b>(text: &'a str, pattern: &'b str) -> bool {
    text.ends_with(pattern)
}
#[doc = "Case-insensitive pattern matching"]
#[doc = " Depyler: verified panic-free"]
#[doc = " Depyler: proven to terminate"]
pub fn case_insensitive_match<'b, 'a>(text: &'a str, pattern: &'b str) -> bool {
    let text_lower: String = text.to_lowercase();
    let pattern_lower: String = pattern.to_lowercase();
    let _cse_temp_0 = text_lower.contains(pattern_lower);
    let matches: bool = _cse_temp_0;
    matches
}
#[doc = "Find text between two markers"]
#[doc = " Depyler: verified panic-free"]
#[doc = " Depyler: proven to terminate"]
pub fn find_between<'a, 'c, 'b>(
    text: &'a str,
    start_marker: &'b str,
    end_marker: &'c str,
) -> String {
    let mut start_pos: i32 = text.find(start_marker).map(|i| i as i32).unwrap_or(-1);
    let _cse_temp_0 = start_pos < 0;
    if _cse_temp_0 {
        return STR_EMPTY;
    }
    let _cse_temp_1 = start_marker.len() as i32;
    start_pos = format!("{}{}", start_pos, _cse_temp_1);
    let end_pos: i32 = text[start_pos as usize..]
        .find(end_marker)
        .map(|i| (i + start_pos as usize) as i32)
        .unwrap_or(-1);
    let _cse_temp_2 = end_pos < 0;
    if _cse_temp_2 {
        return STR_EMPTY;
    }
    let mut result: String = {
        let base = text;
        let start_idx: i32 = start_pos;
        let stop_idx: i32 = end_pos;
        let len = base.chars().count() as i32;
        let actual_start = if start_idx < 0 {
            (len + start_idx).max(0) as usize
        } else {
            start_idx.min(len) as usize
        };
        let actual_stop = if stop_idx < 0 {
            (len + stop_idx).max(0) as usize
        } else {
            stop_idx.min(len) as usize
        };
        if actual_start < actual_stop {
            base.chars()
                .skip(actual_start)
                .take(actual_stop - actual_start)
                .collect::<String>()
        } else {
            String::new()
        }
    };
    result
}
#[doc = "Replace multiple patterns"]
pub fn replace_multiple<'b, 'a>(
    text: &'a str,
    replacements: &'b Vec<()>,
) -> Result<String, IndexError> {
    let mut result: String = text;
    for replacement in replacements.iter().cloned() {
        let old: String = replacement.get(0usize).cloned().unwrap_or_default();
        let new: String = replacement.get(1usize).cloned().unwrap_or_default();
        result = result.replace(old, new);
    }
    Ok(result)
}
#[doc = "Count occurrences of a word"]
#[doc = " Depyler: verified panic-free"]
pub fn count_word_occurrences<'a, 'b>(text: &'a str, word: &'b str) -> i32 {
    let words: Vec<String> = text
        .split_whitespace()
        .map(|s| s.to_string())
        .collect::<Vec<String>>();
    let mut count: i32 = 0;
    for w in words.iter().cloned() {
        if w == word {
            count = count + 1;
        }
    }
    count
}
#[doc = "Extract numbers from text"]
#[doc = " Depyler: verified panic-free"]
pub fn extract_numbers_from_text(text: &str) -> Vec<i32> {
    let mut numbers: Vec<i32> = vec![];
    let mut current_num: String = STR_EMPTY.to_string();
    for _char in text.chars() {
        let char = _char.to_string();
        let mut current_num;
        if char.chars().all(|c| c.is_numeric()) {
            current_num = current_num + char;
        } else {
            if current_num.len() as i32 > 0 {
                let mut num: i32 = (current_num) as i32;
                numbers.push(num);
                current_num = STR_EMPTY;
            }
        }
    }
    let _cse_temp_0 = current_num.len() as i32;
    let _cse_temp_1 = _cse_temp_0 > 0;
    if _cse_temp_1 {
        let _cse_temp_2 = (current_num) as i32;
        let mut num: i32 = _cse_temp_2;
        numbers.push(num);
    }
    numbers
}
#[doc = "Simple wildcard matching(* means any sequence)"]
#[doc = " Depyler: proven to terminate"]
pub fn wildcard_match_simple<'a, 'b>(text: &'a str, pattern: &'b str) -> Result<bool, IndexError> {
    let _cse_temp_0 = !pattern.contains(&"*");
    if _cse_temp_0 {
        return Ok(text == pattern);
    }
    let parts: Vec<String> = pattern
        .split("*")
        .map(|s| s.to_string())
        .collect::<Vec<String>>();
    let _cse_temp_1 = parts.len() as i32;
    let _cse_temp_2 = _cse_temp_1 != 2;
    if _cse_temp_2 {
        return Ok(false);
    }
    let prefix: String = parts.get(0usize).cloned().unwrap_or_default();
    let suffix: String = parts.get(1usize).cloned().unwrap_or_default();
    let mut has_prefix: bool = true;
    let mut has_suffix: bool = true;
    let _cse_temp_3 = prefix.len() as i32;
    let _cse_temp_4 = _cse_temp_3 > 0;
    if _cse_temp_4 {
        has_prefix = text.starts_with(prefix);
    }
    let _cse_temp_5 = suffix.len() as i32;
    let _cse_temp_6 = _cse_temp_5 > 0;
    if _cse_temp_6 {
        has_suffix = text.ends_with(suffix);
    }
    Ok((has_prefix) && (has_suffix))
}
#[doc = "Run all regex module tests"]
#[doc = " Depyler: verified panic-free"]
#[doc = " Depyler: proven to terminate"]
pub fn test_all_re_features() -> Result<(), Box<dyn std::error::Error>> {
    let matches: bool = test_simple_match();
    let contains: bool = test_contains_pattern();
    let position: i32 = test_find_pattern_position();
    let mut count: i32 = test_count_occurrences();
    let replaced: String = test_replace_pattern();
    let split_result: Vec<String> = test_split_by_pattern();
    let is_digit: bool = test_match_digit();
    let is_alpha: bool = test_match_alpha();
    let is_alnum: bool = test_match_alphanumeric();
    let text: String = "abc123def456".to_string();
    let mut digits: String = extract_digits(text);
    let mut letters: String = extract_letters(text);
    let sentence: String = "Hello world from Python".to_string();
    let words: Vec<String> = find_all_words(sentence);
    let email_valid: bool = validate_email_simple("user@example.com");
    let email_invalid: bool = validate_email_simple("notanemail");
    let phone_valid: bool = validate_phone_simple("555-123-4567");
    let phone_invalid: bool = validate_phone_simple("abc");
    let mut url: String = "https://www.example.com/path/page.html".to_string();
    let mut domain: String = extract_url_domain(url);
    let punct_text: String = "Hello, World!".to_string();
    let no_punct: String = remove_punctuation(punct_text);
    let spaces: String = "Hello    World  !".to_string();
    let normalized: String = normalize_whitespace(spaces);
    let starts: bool = starts_with_pattern(STR_HELLO_WORLD, STR_HELLO);
    let ends: bool = ends_with_pattern(STR_HELLO_WORLD, "World");
    let case_match: bool = case_insensitive_match(STR_HELLO, "hello");
    let tagged: String = "<tag>content</tag>".to_string();
    let content: String = find_between(tagged, "<tag>", "</tag>");
    let replacements: Vec<()> = vec![("a", "x"), ("b", "y")];
    let multi_replace: String = replace_multiple("aabbcc", &replacements);
    let para: String = "the quick brown fox jumps over the lazy dog".to_string();
    let the_count: i32 = count_word_occurrences(para, "the");
    let mixed: String = "I have 2 apples and 5 oranges".to_string();
    let nums: Vec<i32> = extract_numbers_from_text(mixed);
    let wildcard1: bool = wildcard_match_simple("hello.txt", "*.txt");
    let wildcard2: bool = wildcard_match_simple("test_file.py", "test_*");
    println!("{}", "All regex module tests completed successfully");
    Ok(())
}
#[cfg(test)]
mod tests {
    use super::*;
    use quickcheck::{quickcheck, TestResult};
    #[test]
    fn test_test_find_pattern_position_examples() {
        let _ = test_find_pattern_position();
    }
    #[test]
    fn test_test_count_occurrences_examples() {
        let _ = test_count_occurrences();
    }
    #[test]
    fn test_validate_email_simple_examples() {
        let _ = validate_email_simple(Default::default());
    }
    #[test]
    fn test_validate_phone_simple_examples() {
        let _ = validate_phone_simple(Default::default());
    }
    #[test]
    fn quickcheck_normalize_whitespace() {
        fn prop(text: String) -> TestResult {
            let once = normalize_whitespace((&*text).into());
            let twice = normalize_whitespace(once.clone());
            if once != twice {
                return TestResult::failed();
            }
            TestResult::passed()
        }
        quickcheck(prop as fn(String) -> TestResult);
    }
}
